Update Taints and Tolerations Documentation

network-charles opened the pull request:
# Lack of Clarity in the Original 'Taints and Tolerations' Documentation

* It didn't let developers know about another feature that can override a taint configuration.

# Conclusion

Developers should be mindful that manually scheduling pods with node names in YAML spec files can override standard scheduling mechanisms and bypass taint configurations, impacting deployment behaviour in their Kubernetes cluster.

Thanks to @windsonsea and @tengqm who helped me align my contribution.

stmcginnis commented:
/lgtm

windsonsea commented on content/en/docs/concepts/scheduling-eviction/taint-and-toleration.md:
```diff
@@ -315,6 +315,8 @@ tolerations to all daemons, to prevent DaemonSets from breaking.
 Adding these tolerations ensures backward compatibility. You can also add
 arbitrary tolerations to DaemonSets.
 
+> **Note:** If you manually schedule a pod by defining a node name in your workload yaml spec file like a pod, deployments, etc. it will override the usual scheduling mechanism, hence, bypassing the taint.
```
```suggestion
{{< note >}}
If you manually schedule a pod by defining a node name in your workload yaml spec
file like a pod, deployments, etc. it will override the usual scheduling mechanism,
hence, bypassing the taint.
{{< /note >}}
```

kundan2707 commented:
/lgtm

kundan2707 commented:
/lgtm

network-charles commented:
This is my first contribution to the repo. I’m not sure if I need to wait for all the reviewers, but the bot mentioned that with an LGTM label, I can assign Diva for approval, so I’m following that guidance.

/assign @divya-mohan0209

tengqm commented on content/en/docs/concepts/scheduling-eviction/taint-and-toleration.md:
```diff
@@ -315,6 +315,12 @@ tolerations to all daemons, to prevent DaemonSets from breaking.
+{{< note >}}
+If you manually schedule a pod by defining a node name in your workload yaml spec
+file like a pod, deployments, etc. it will override the usual scheduling mechanism,
+hence, bypassing the taint.
```
There are a few nits in this note. For example, we don't say "workload yaml spec file",
we say "workload manifest". "... like a pod, deployments" are inaccurate. When mentioning a resource kind, use the resource names is preferred, i.e. "Pod" and "Deployment". The scheduling mechanism is not overridden. I don't think this claim is correct. If the `nodeName` for a Pod manifest is manually specified, the scheduler will assign the Pod to the node specified, regardless of any taints or tolerations. I think this is what you were trying to clarify.

network-charles replied:
Nice, thanks for the review @tengqm. Yes, that was what I was trying to clarify. What do you think about the correction now?

>{{< note >}}
If you manually schedule a pod by defining a `nodeName` in your workload manifest like a "Pod" and "Deployment", the scheduler will assign the Pod to the node specified, regardless of any taints or tolerations.
{{< /note >}}

Arhell reviewed (commented):
lgtm

sftim commented on content/en/docs/concepts/scheduling-eviction/taint-and-toleration.md:
```diff
@@ -315,6 +315,10 @@ tolerations to all daemons, to prevent DaemonSets from breaking.
 arbitrary tolerations to DaemonSets.
 
+{{< note >}}
+If the `nodeName` for a Pod manifest is manually specified, the scheduler will assign the Pod to the node specified, regardless of any taints or tolerations.
```
If you reword this, I recommend you [write in the active voice](https://kubernetes.io/docs/contribute/style/style-guide/#use-active-voice).

network-charles replied:
Lovely, thanks for the review @sftim. I will modify it to the below:
```markdown
{{< note >}}
If you manually specify the `nodeName` for a Pod manifest, the scheduler will assign the Pod to the specified node, regardless of any taints or tolerations.
{{< /note >}}
```

I also want to suggest that it appears under the heading "Concepts". The example in the heading was what sprung the PR in the first place. What do you think?

sftim replied:
You're still using a `note` (I wouldn't), and you're still mentioning the scheduler assigning the Pod to a node (which in this case is not going to happen; the scheduler only ever binds _unbound_ Pods to nodes).

The move sounds OK.

network-charles replied:
I get it now. 
If that's so, adding it just above the statement "Here's an example of a pod that uses tolerations:" without a note makes sense because developers have just learned how to prevent pods from being scheduled to inappropriate nodes using taints and tolerations. However, they may be unaware of how a previous factor, `nodeName`, which they learned earlier in the [Assigning Pods to Nodes](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/) page can influence this new configuration.

The final statement becomes:
**However, if you manually specify the `nodeName` for a Pod manifest without specifying a matching toleration, the kubelet on the named tainted node tries to place the Pod on that node, regardless of any taints or tolerations.**
 
**Here's an example of a pod that uses tolerations:**

What do you think of the final statement and also where it will be newly placed compared to after the section 'Concept'?
In the final statement, I also thought about using "in the PodSpec" instead of "for a Pod manifest" since PodSpec was used on the page.

sftim replied:
Try this:
```markdown
The default Kubernetes scheduler takes taints and tolerations into account when
selecting a node to run a particular Pod. However, if you manually specify the
`.spec.nodeName` for a Pod, that action bypasses the scheduler; the Pod is then
bound onto on the node where you assigned it, even if there are `NoSchedule`
taints on that node that you selected.
If this happens and the node also has a `NoExecute` taint set, the kubelet will
eject the Pod unless there is an appropriate toleration set.

Here's an example of a pod that has some tolerations defined:
```

network-charles replied:
I appreciate your review @sftim, thank you!

reylejano commented:
/approve

@network-charles , the description leads me to think there is a related issue. In the description, please add a link to the issue if there is a related issue

network-charles commented:
> /approve
> 
> @network-charles , the description leads me to think there is a related issue. In the description, please add a link to the issue if there is a related issue

Hi @reylejano, there is no related issue. I’ve updated the comment and replaced the word “issue” with “lack of clarity” instead.

ricardoamaro commented on content/en/docs/concepts/scheduling-eviction/taint-and-toleration.md:
```diff
@@ -62,7 +62,15 @@ tolerations:
+bound onto on the node where you assigned it, even if there are `NoSchedule`
+taints on that node that you selected.
+If this happens and the node also has a `NoExecute` taint set, the kubelet will
+eject the Pod unless there is an appropriate toleration set.
```
```suggestion
eject the Pod unless there is an appropriate tolerance set.
```

ricardoamaro commented on content/en/docs/concepts/scheduling-eviction/taint-and-toleration.md:
```diff
@@ -62,7 +62,15 @@ tolerations:
+The default Kubernetes scheduler takes taints and tolerations into account when
+selecting a node to run a particular Pod. However, if you manually specify the
+`.spec.nodeName` for a Pod, that action bypasses the scheduler; the Pod is then
+bound onto on the node where you assigned it, even if there are `NoSchedule`
```
```suggestion
bound onto the node where you assigned it, even if there are `NoSchedule`
```

network-charles commented:
> /approve
> 
> @network-charles , the description leads me to think there is a related issue. In the description, please add a link to the issue if there is a related issue

Hi Rey, the merging is still blocked even though issue #45890 is closed.

windsonsea commented:
/lgtm
Thanks.

windsonsea commented:
/label tide/merge-method-merge
